src/pointer/transmute.rs

```diff
@@ -269,75 +267,133 @@ where
+///   `size_of::<Dst>()` bytes of any `SV`-valid `Src` must be a `DV`-valid
+///   `Dst`.
+/// - If `size_of::<Src>() < size_of::<Dst>()`, then any `SV`-valid `Src`
+///   followed by `size_of::<Dst>() - size_of::<Src>()` uninitialized bytes must
+///   be a `DV`-valid `Dst`.
+///
+/// If either `Src: !Sized` or `Self: !Sized`, then this condition does not need
+/// to hold.
+///
+/// ## By-reference transmutation
+///
+/// If `Dst: SizeCompat<Src>`, then the following must hold: For all [valid
+/// sizes] of `Src`, `ssize` let `s: NonNull<Src>` with referent size `ssize`.
+/// Let `dsize` be the referent size of `SizeCompat::cast_from_raw(s)`. Note
+/// that, by invariant on `cast_from_raw`, `ssize >= dsize`. For all `SV`-valid
+/// values of `Src` with size `ssize`, `src`, it must be the case that the
+/// leading `dsize` bytes of `src` constitute a `DV`-valid `Dst`.
+///
+/// *This case corresponds to a `&Src` to `&Dst` transmute.*
+///
+/// If `Src: SizeCompat<Dst>`, then the following must hold: For all [valid
+/// sizes] of `Dst`, `dsize`, let `d: NonNull<Dst>` with referent size `dsize`.
+/// Let `ssize` be the referent size of `SizeCompat::cast_from_raw(d)`. Note
+/// that, by invariant on `cast_from_raw`, `dsize >= ssize`. For all `DV`-valid
+/// values of `Dst` with size `dsize`, `dst`, and for all `SV`-valid values of
+/// `Src` with size `ssize`, `src`, let `dst'` be constructed by concatenating
+/// `src` with the trailing `dsize - ssize` bytes of `dst`. It must be the case
+/// that `dst'` is a `DV`-valid `Dst`.
+///
+/// *This case corresponds to a `&Dst` to `&Src` transmute where `Src` and `Dst`
+/// permit interior mutation, or to a `&mut Dst` to `&mut Src` transmute. In
+/// particular, it ensures that, once values have been written to the `&Src` or
+/// `&mut Src` and the `&Src` or `&mut Src` have been dropped, the original
+/// `&Src` or `&mut Src` still refer to an `SV`-valid `Src`.*
+///
+/// Note that, if `<Dst as SizeCompat<Src>>::cast_from_raw` and `<Src as
+/// SizeCompat<Dst>>::cast_from_raw` both preserve referent size exactly, then
+/// the conditions in this section are a logical consequence of the conditions
+/// in the "By-value transmutation" section.
+///
+/// [valid sizes]: crate::KnownLayout#what-is-a-valid-size
+pub unsafe trait TransmuteFrom<Src: ?Sized, SV, DV> {}
 
 /// # Safety
 ///
-/// `T` and `Self` must have the same vtable kind (`Sized`, slice DST, `dyn`,
-/// etc) and have the same size. In particular:
-/// - If `T: Sized` and `Self: Sized`, then their sizes must be equal
-/// - If `T: ?Sized` and `Self: ?Sized`, then it must be the case that, given
-///   any `t: *mut T`, `t as *mut Self` produces a pointer which addresses the
-///   same number of bytes as `t`.
-pub unsafe trait SizeEq<T: ?Sized> {
-    fn cast_from_raw(t: NonNull<T>) -> NonNull<Self>;
+/// Implementations of `cast_from_raw` must satisfy that method's safety
+/// post-condition.
+pub unsafe trait SizeCompat<Src: ?Sized> {
```

Comment:
We don't need to do this *now*, but it'd be more idiomatic to name this `CastFrom`, after its method (and source of safety conditions). That name also conveys directionality, unlike `SizeCompat`.